[Gutenberg] Add new block type badge to block inserter list

dcalhoun opened the pull request:
* **Gutenberg:** https://github.com/WordPress/gutenberg/pull/33119
* **Gutenberg Mobile:** https://github.com/wordpress-mobile/gutenberg-mobile/pull/3720
* **WPiOS:** https://github.com/wordpress-mobile/WordPress-iOS/pull/16857
* **WPAndroid:** Self

Fixes WordPress/gutenberg#30627. Add badge to showcase new block types. 

To test: See https://github.com/WordPress/gutenberg/pull/33119. 

## Regression Notes
1. Potential unintended areas of impact
  Modifications to the Gutenberg view and settings could cause disruption initializing the editor. 
2. What I did to test those areas of impact (or what existing automated tests I relied on)
  Verified the editor loaded without issue and successfully modified/saved post content. 
3. What automated tests I added (or what prevented me from doing so)
  None, as this feature is experimental at this time and any tests would be implemented in the Gutenberg editor repo.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dcalhoun committed:
Update gutenbergMobileVersion ref

dcalhoun committed:
[wip] Explore storage of block type impressions

dcalhoun committed:
Refactor Gutenberg bridge methods to match new implementation

The Gutenberg bridge methods were refactored to avoid "leaking" React
Native types deeper into WPAndroid's source. These changes update the
methods to match those new signatures.

dcalhoun committed:
Store and retrieve block type impressions from AppPrefs

We track changes to the block type impressions to appropriate remove the
"new" badge for a give block type.

dcalhoun committed:
Remove lingering to-do comment

dcalhoun committed:
Address lint errors

Remove unused imports and shorten line length.

dcalhoun committed:
Split tracking of editor onboarding capability and initial session

Previously, the `editorOnboarding` capability included both (1) whether
the user was within the cohort and (2) if the user had previously
launched the Gutenberg editor. The latter is utilized to only show the
initial tooltip once. However, the combination resulted in the latter
preventing the new block badges from displaying past the initial
session.

Splitting these two attributes allows for features like the new blocks
badge to remain active past the user's initial editor session.

dcalhoun committed:
Update gutenbergMobileVersion ref

mchowning commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -125,6 +130,7 @@
         SHOULD_AUTO_ENABLE_GUTENBERG_FOR_THE_NEW_POSTS_PHASE_2,
         GUTENBERG_OPT_IN_DIALOG_SHOWN,
         GUTENBERG_FOCAL_POINT_PICKER_TOOLTIP_SHOWN,
+        GUTENBERG_BLOCK_TYPE_IMPRESSIONS,
```
Because this is a `DeleteablePrefKey` this will get reset if the user logs in and logs out. This is a bit of an edge case, but a user who regularly switches accounts would have the impressions reset each time they switched. I wonder if we would be better off making this an `UndeleteablePrefKey` so the preferences would persist even if they switched accounts.

dcalhoun replied:
@mchowning good thought. Persisting the value throughout the life of the app install was my original intent. I agree that resetting this value when a user logs out would likely be unexpected. I will update the value to be an `UndeleteablePrefKey`.

dcalhoun replied:
Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/15038/commits/03dee0565ec8325a0f515dd94be5a298f00badd9.

dcalhoun committed:
Remove logic that is now implemented in JavaScript

To de-duplicate and simplify bridging code, more logic was relocated to
the JavaScript for the React Native app.

dcalhoun committed:
Relocate impressions key to UndeleteablePrefKey

The block type impression counts should not reset when a user logs out.

dcalhoun committed:
Update gutenbergMobileVersion ref

dcalhoun committed:
Update gutenbergMobileVersion ref

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into add/native-new-block-badge

mchowning commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/EditorFragmentAbstract.java:
```diff
@@ -219,6 +219,8 @@ public static MediaType getEditorMimeType(MediaFile mediaFile) {
         void onReplaceStoryEditedBlockActionReceived();
         boolean showPreview();
+        Map<String, Double> onRequestBlockTypeImpressions();
+        void onSetBlockTypeImpressions(Map<String, Double> impressions);
```
Just FYI: Instead of redeclaring these methods, we could have this interface extend the `OnBlockTypeImpressionsEventListener` interface. Doing that would also allow us to just pass the `editorFragmentListener` as the implementation of the `OnBLockTypeImpressionsEventListener` interface directly instead of creating an anonymous implementation of the interface in `GutenbergEditorFragment::onCreateView`. I created a commit showing what this would look like [here](https://github.com/wordpress-mobile/WordPress-Android/commit/32265822f6a020e6b44f5c8fdbeb5267337e06eb).

I think this is a nice change, but no strong feelings. Definitely not a blocker, so only make this change if you like it better. 🙂

dcalhoun replied:
Interesting. Thank you for sharing this approach. In particular, thank you for sharing the example diff, as it was difficult for me to follow the written description. 

As a novice Android developer, I crafted the current implementation following the preexisting patterns in the code base for similar features. However, I see how leveraging extensions could make this less verbose. 

That said, I would prefer to forego the change at this time, as it appears it would require additional changes to the Gutenberg Mobile Java interface as well. I'd prefer to keep that stable now that it has been approved and I hope to merge later today.

mchowning reviewed (approved):
Looks great! Nice work @dcalhoun ! 🎉

dcalhoun committed:
Update gutenbergMobileVersion ref

dcalhoun committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into add/native-new-block-badge

dcalhoun committed:
Update gutenbergMobileVersion to v1.59.0-alpha2 tag

dcalhoun committed:
Update gutenbergMobileVersion ref

Leverage <branch>-<commit SHA> pattern as the tag appeared to cause
failures with buildkite.

dcalhoun committed:
Revert "Update gutenbergMobileVersion ref"

This reverts commit ba50b8f93f47d0ec432feed8c766fda3c7a97b95.
